source/hpack/tables.d

```diff
@@ -0,0 +1,344 @@
+//module vibe.http.internal.hpack.tables;
+module HPACK.tables;
+
+import vibe.http.status;
+import vibe.http.common;
+import vibe.http.internal.http2;
+import vibe.core.log;
+import vibe.internal.array : FixedRingBuffer;
+
+import std.variant;
+import std.traits;
+import std.meta;
+import std.range;
+import taggedalgebraic;
+
+/*
+	2.3.  Indexing Tables
+
+	HPACK uses two tables for associating header fields to indexes.  The
+	static table (see Section 2.3.1) is predefined and contains common
+	header fields (most of them with an empty value).  The dynamic table
+	(see Section 2.3.2) is dynamic and can be used by the encoder to
+	index header fields repeated in the encoded header lists.
+
+	These two tables are combined into a single address space for
+	defining index values (see Section 2.3.3).
+
+ 2.3.1.  Static Table
+
+	The static table consists of a predefined static list of header
+	fields.  Its entries are defined in Appendix A.
+
+ 2.3.2.  Dynamic Table
+
+	The dynamic table consists of a list of header fields maintained in
+	first-in, first-out order.  The first and newest entry in a dynamic
+	table is at the lowest index, and the oldest entry of a dynamic tabl
+	is at the highest index.
+	The dynamic table is initially empty.  Entries are added as each
+	header block is decompressed.
+
+	The dynamic table is initially empty.  Entries are added as each
+	header block is decompressed.
+
+	The dynamic table can contain duplicate entries (i.e., entries with
+	the same name and same value).  Therefore, duplicate entries MUST NOT
+	be treated as an error by a decoder.
+
+	The encoder decides how to update the dynamic table and as such can
+	control how much memory is used by the dynamic table.  To limit the
+	memory requirements of the decoder, the dynamic table size is
+	strictly bounded (see Section 4.2).
+
+	The decoder updates the dynamic table during the processing of a list
+	of header field representations (see Section 3.2).
+*/
+
+// wraps a header field = name:value
+struct HTTP2HeaderTableField {
+	private union HeaderValue {
+		string str;
+		string[] strarr;
+		HTTPStatus status;
+		HTTPMethod method;
+	}
+
+	string name;
+	TaggedAlgebraic!HeaderValue value;
+
+	// initializers
+	static foreach(t; __traits(allMembers, HeaderValue)) {
+		mixin("this(string n, " ~
+				typeof(__traits(getMember, HeaderValue, t)).stringof ~
+				" v) { name = n; value = v; }");
+	}
+}
+
+// fixed as per HPACK RFC
+immutable size_t STATIC_TABLE_SIZE = 61;
+
+/** static table to index most common headers
+  * fixed size, fixed order of entries (read only)
+  * cannot be updated while decoding a header block
+  */
+static immutable HTTP2HeaderTableField[STATIC_TABLE_SIZE+1] StaticTable;
+
+static this() {
+	StaticTable = [
```

Comment:
Probably won't matter in practice, but since this is a regular array, the initialization can now be static (`static immutable H2F[N] StaticTable = [...];`), with the advantage that it doesn't have any runtime overhead and doesn't rely on the module constructor mechanism.

Reply:
I made the same considerations but then TaggedAlgebraic complained about it this way:

```D
Error: array cast from void[16] to string[] is not supported at compile time
~/.dub/packages/taggedalgebraic-0.10.11/taggedalgebraic/source called from here: delegate () => cast(string[])dst[0..16]()
called from here: rawEmplace(cast(void[])this.m_data.data, value)
called from here: this.value
```
That's why I decided to use a static initializer (which doesn't make much of a difference anyway, apart from the module constructor being involved).

Reply:
Ah okay, interesting. I'll have to look into making it CTFE-compatible at some point then.